src/librustc_incremental/calculate_svh.rs

```diff
@@ -74,17 +36,24 @@ impl Svh {
         // to ensure it is not incorporating implementation artifacts into
         // the hash that are not otherwise visible.)
 
+        let crate_disambiguator = self.sess.crate_disambiguator.get();
+        let krate = self.map.krate();
+
         // FIXME: this should use SHA1, not SipHash. SipHash is not built to
         //        avoid collisions.
         let mut state = SipHasher::new();
+        debug!("state: {:?}", state);
 
         "crate_disambiguator".hash(&mut state);
-        crate_disambiguator.len().hash(&mut state);
-        crate_disambiguator.hash(&mut state);
+        crate_disambiguator.as_str().len().hash(&mut state);
```

Comment:
On Mon, Apr 04, 2016 at 08:48:39AM -0700, Michael Woerister wrote:
> >  
> >          "crate_disambiguator".hash(&mut state);
> > -        crate_disambiguator.len().hash(&mut state);
> > -        crate_disambiguator.hash(&mut state);
> > +        crate_disambiguator.as_str().len().hash(&mut state);
> 
> I'm not sure if it's a good idea to re-use the SVH for incremental compilation. It's something with a specific purpose (2) and trying to make it hit two goals at once might just make it more complicated. The other way round though, re-using a well-defined incremental compilation hash when computing the SVH makes more sense to me.

I agree that the SVH as originally envisioned is not what we want --
for one thing, it tries to have some (broken) notion of not hashing
things that do not affect "ABI compatibilty". I think there is no use
for the existing SVH. I am basically taking it over and repurposing
it. I should rename it to "ICH" (incr. comp. hash), probably. But the
current names seems also not inappropriate.

Do you think that it's worth retaining the "SVH as it exists today"?

> ad endianess: Yes, for incremental compilation it doesn't make a
> difference. For the other case: I guess you can never have two
> crates with differing endianess in the your crate graph. So, in the
> end it should not matter in either case. It's more of code smell
> than a bug anyway.

well, what I was concerned about is that you cross-compile the library
on one machine. then copy it somewhere else. then recompile something
it depends on on that local machine. now if the endianness's differ
rustc may get whiny, even though same toolchain/version/everything was
used, because the SVH recorded in the (cross-compiled) crate is
different from the one that is found in the (locally) compiled dep.
Admittedly this scenario seems a bit far fetched and like something
that can be relegated to a FIXME.


Niko